hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java

```diff
@@ -470,12 +490,209 @@ protected static List<OzoneAcl> getAclsForDir(KeyArgs keyArgs,
+    String bucketName = keyArgs.getBucketName();
+    String keyName = keyArgs.getKeyName();
+
+    for (String missingKey : missingParents) {
+      long nextObjId = baseObjId + objectCount;
+      if (nextObjId > maxObjId) {
+        throw new OMException("Too many directories in path. Exceeds limit of "
+            + MAX_NUM_OF_RECURSIVE_DIRS + ". Unable to create directory: "
+            + keyName + " in volume/bucket: " + volumeName + "/" + bucketName,
+            INVALID_KEY_NAME);
+      }
+
+      LOG.debug("missing parent {} getting added to KeyTable", missingKey);
+
+      OmKeyInfo parentKeyInfo =
+          createDirectoryKeyInfoWithNoACL(missingKey, keyArgs, nextObjId,
+              bucketInfo, omPathInfo, trxnLogIndex,
+              ozoneManager.getDefaultReplicationConfig(), ozoneManager.getConfiguration());
+      objectCount++;
+
+      missingParentInfos.add(parentKeyInfo);
+    }
+
+    return missingParentInfos;
+  }
+
+  /**
+   * Fill in a DirectoryInfo for a new directory entry in OM database.
+   * @param dirName
+   * @param keyArgs
+   * @param objectId
+   * @param parentObjectId
+   * @param bucketInfo
+   * @param omPathInfo
+   * @return the OmDirectoryInfo structure
+   */
+  @SuppressWarnings("parameternumber")
+  protected OmDirectoryInfo createDirectoryInfoWithACL(
+      String dirName, KeyArgs keyArgs, long objectId,
+      long parentObjectId, long transactionIndex,
+      OmBucketInfo bucketInfo, OMFileRequest.OMPathInfo omPathInfo,
+      OzoneConfiguration config) throws OMException {
+    return OmDirectoryInfo.newBuilder()
+        .setName(dirName)
+        .setOwner(keyArgs.getOwnerName())
+        .setCreationTime(keyArgs.getModificationTime())
+        .setModificationTime(keyArgs.getModificationTime())
+        .setObjectID(objectId)
+        .setUpdateID(transactionIndex)
+        .setParentObjectID(parentObjectId).setAcls(getAclsForDir(keyArgs, bucketInfo, omPathInfo, config))
+        .build();
+  }
+
+  /**
+   * fill in a KeyInfo for a new directory entry in OM database.
+   * without initializing ACLs from the KeyArgs - used for intermediate
+   * directories which get created internally/recursively during file
+   * and directory create.
+   * @param keyName
+   * @param keyArgs
+   * @param objectId
+   * @param bucketInfo
+   * @param omPathInfo
+   * @param transactionIndex
+   * @param serverDefaultReplConfig
+   * @return the OmKeyInfo structure
+   */
+  @SuppressWarnings("parameternumber")
+  protected OmKeyInfo createDirectoryKeyInfoWithNoACL(String keyName,
+      KeyArgs keyArgs, long objectId, OmBucketInfo bucketInfo,
+      OMFileRequest.OMPathInfo omPathInfo, long transactionIndex,
+      ReplicationConfig serverDefaultReplConfig, OzoneConfiguration config) throws OMException {
+    return dirKeyInfoBuilderNoACL(keyName, keyArgs, objectId,
+        serverDefaultReplConfig)
+        .setAcls(getAclsForDir(keyArgs, bucketInfo, omPathInfo, config))
```

Comment:
Do we need `setAcls` here?

Reply:
setAcls is needed. It's original name is createDirectoryKeyInfoWithACL. I rollback the change to use its original name.